### PR TITLE
[JENKINS-50394] Get remote references before fetching to cache when discoverX

### PR DESCRIPTION
**Pull request by rsandell**
## [JENKINS-50394](https://issues.jenkins-ci.org/browse/[JENKINS-50394) - Get remote references before fetching to local cache when discovering branches et.al.

The MissingObjectException thrown is due to that the local cache is first fetched to and then remote references are used to list branches etc. So if the repository is somewhat large, so the fetch takes some time, and new commits are pushed during that fetch operation; the later listing of remote refs might discover branches pointing to objects that aren't in the local cache.
This fixes that by listing the remote references first and then doing the fetch.
Some situations might still cause the issue though like a rebase or a force push, but that should be less likely since listing the remote references is fairly quick even on large repositories.

To be able to test this a small change was needed in git-client https://github.com/jenkinsci/git-client-plugin/pull/420

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [X] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [X] I have referenced the Jira issue related to my changes in one or more commit messages
- [X] I have added tests that verify my changes
- [X] Unit tests pass locally with my changes
- [X] I have added documentation as necessary
- [X] No Javadoc warnings were introduced with my changes
- [X] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [X] Bug fix (non-breaking change which fixes an issue)

## Further comments

I was at first hoping to be able to do even more lazy fetching to the local cache so that not all repository scans would need it. But the way branch-api assumes things made that extremely hard to do in a reliable way.

### CODE REVIEW

**Comment by rsandell** [comment]
@MarkEWaite @stephenc @bitwiseman @dwnusbaum @abayer @kshultzCB @car-roll

**Comment by rsandell** [comment]
The target branch doesn't have incrementals enabled and I can't remember how we did it in the past with dependant PRs :)

File: src/test/java/jenkins/plugins/git/AbstractGitSCMSourceTest.java [hunk 1]
```diff
@@ -864,6 +888,126 @@ public void testCustomRefSpecs() throws Exception {
         assertEquals("+refs/heads/*:refs/remotes/origin/* +refs/merge-requests/*/head:refs/remotes/origin/merge-requests/*", config.getRefspec());
     }
 
+    @Test @Issue("JENKINS-50394")
```
Comment (rsandell) [inline src/test/java/jenkins/plugins/git/AbstractGitSCMSourceTest.java]: I wrote the test before implementing the fix to make sure I could recreate it.

**Comment by rsandell** [comment]
@MarkEWaite Anything else you need from me to get this merged and released?

**Comment by MarkEWaite** [comment]
@rsandell could you also propose a matching pull request that would merge these same changes to the master branch?

I need time to perform some interactive testing of the change before I merge and release it.

**Comment by rsandell** [comment]
@MarkEWaite done, see #706

**Comment by MarkEWaite** [comment]
Thanks @rsandell !  I completed my interactive testing and have merged this change.  Will likely create a new release of the git plugin soon so that the community can use the fix.

@Dohbedoh would you be willing to experiment with it as well?

**Pull request merged**